### DISCUSSION TITLE
How to buffer data into MEMTILE and read it multiple Times?

### DISCUSSION THREAD

**Discussion by ngdymx** [Xilinx/mlir-aie#1566] [Q&A] [answered]
Hi team,

Is there any way that I can buffer data into MEMTILE and read it multiple times?
Thanks a lot!

**Comment by AndraBisca** [comment, marked as answer]
To do that, you can set the [repeat_count](https://github.com/Xilinx/mlir-aie/blob/main/include/aie/Dialect/AIE/IR/AIEOps.td#L934) input of the AIE_DMAStartOp to a value >= 1, depending on how many **more** times you will read the data. This will enable the DMA associated with your data movement to push a new copy of the data on the stream.

I am currently working on adding a feature to the AIE_ObjectFifoCreateOp to be able to describe this data repeat, however, the [PR](https://github.com/Xilinx/mlir-aie/pull/1530) isn't ready just yet. I am aware of some examples without object FIFOs that currently use the repeat feature [here](https://github.com/Xilinx/mlir-aie/blob/main/test/npu-xrt/e2e/test_repeat_count.py).

Hope this helps!

**Reply by ngdymx** [reply]
Hi,

Thank you so much! I will try it.

**Reply by ngdymx** [reply]
Hi,

I hope you are doing well.

I wanted to inquire about the current status of adding repeat count to objectFifo. I understand you might be busy, and I do not mean to rush you. However, I wanted to let you know that I am still awaiting it.

Thank you for your time and understanding.

**Reply by AndraBisca** [reply]
Hello,

Thank you for your patience. Unfortunately, while we have made some progress with testing this feature, it has been a little slow due to other ongoing work on different features and fixes. However, I would be curious to know whether you've run any tests related to buffering in the memtile without access to the feature for now? Any additional input is welcome to help speed up its implementation.

Thank you for your understanding and also for checking in on this discussion so that it stays up to date!

**Reply by ngdymx** [reply]
Hello,

I tried to run the ```test_repeat_count.py```, but it gave me the ```cannot import name 'linalg' from 'aie.extras.dialects.ext'``` error. Later, I tried to mimic the code but it failed. ☹️

Then I tried to buffer the data into DDR and read it multiple time, showing in the below code. It worked only if I put "width" in the fourth dimension. If I put the "width" in the second or third dimension, it will fail.

```
npu_dma_memcpy_nd(
    metadata="inOF_wts_0_L3L2",
    bd_id=1,
    mem=W,
    offsets = [0, 0, 0, 0],
    sizes=[width, 1, 1, weightsInInt32s],
    strides=[0, 0, 0],
 )
```

So far, I haven't successfully buffered the data in the MEMTILE.
